### DISCUSSION TITLE
Symfony Runtime - how to prevent access?

### DISCUSSION THREAD

**Discussion by freiondrej-lmc** [symfony/symfony#46722] [Q&A] [answered]
Hello,
we have app.php and app_dev.php entrypoints to our Symfony app. The app_dev.php checks whether the request is trusted based on several conditions and if so, it launches the kernel. If not trusted, it generates a 302 response to app.php.

When we migrated to symfony/runtime, it felt a bit cumbersome in this regard. We ended up creating a dedicated runner extending `\Symfony\Component\Runtime\Runner\Symfony\HttpKernelRunner`, which wraps the Kernel and is returned from the closure in app_dev.php:
```php
// app_dev.php
return function (){
   return new ProtectedRunner($kernel, Request::createFromGlobals());
};
```

Inside the runner's `run()` method, we perform the checks and in case the request is considered non-authorized, we do the following:

```php
// ProtectedRunner.php
return (new ResponseRunner(
          new RedirectResponse('some homepage')
      ))->run();
```

However, it feels weird to instantiate a different runner inside of our ProtectedRunner ... we thought maybe if either ResponseRunner or a RuntimeInterface service was registered in the container, we could do something like this:
```php
// ProtectedRunner.php
return $this->runtime->getRunner(
          new RedirectResponse('some homepage')
      )->run();
```
Even that does not feel right, but would be better than the first example. However, we found no such service in the container, so this is not an option.

Is there a proper way to do this? Or are we bending some symfony principles to a weird shape? Should we perhaps just use firewalls and let the Kernel boot?

@Nyholm if you have the time, could you give us some insight (being the most important Runtime contributor)?

Thank you! :)

**Comment by Nyholm** [comment, marked as answer]
Hey, 
Doing `app.php` and `app_dev.php` is what Symfony recommended way before we started to use environment variables. Is there a reason you dont use environment variables to specify the symfony environment?

**Reply by freiondrej-lmc** [reply]
@Nyholm the reason is probably just historical as of this moment. But even then, the described problem would still apply, or am I missing something? :)

**Reply by Nyholm** [reply]
No, not really. Since you never have "unauthorised access". If you deploy your app with APP_ENV=prod, then the app will not load any of the dev resources. 
Ie, you use different dependency container configs.

The only way to an attacker to see the development things is for them to change your environment variables. That is something the attacker cant on a normal secured system. If they can change/read environment variables, then they will also be able to expose all kind of nasty things (ie, you have bigger problems)

**Reply by freiondrej-lmc** [reply]
Hmm, makes sense 👍 would that mean we cannot activate debug mode on prod anymore? With the current approach, we can load app_dev.php on prod and ProtectedRunner makes sure only our VPN IP is allowed there - is that bad practice?

To my original question, speaking more generally: is there a way for a runner to invoke another runner (as shown in the snippet
```php
// ProtectedRunner.php
return (new ResponseRunner(
          new RedirectResponse('some homepage')
      ))->run();
```
), or is that "too late" once `\Symfony\Component\Runtime\SymfonyRuntime::getRunner()` is called and any possible logic should be executed before returning in https://github.com/symfony/demo/blob/c307b2ad6aa992c1a26d18e9b0b699508dcf4e4c/public/index.php#L8?

**Reply by Nyholm** [reply]
> would that mean we cannot activate debug mode on prod anymore? 

Yes, sorry. 

> is that bad practice?

Hm.. I would use Blackfire, elastic stack or something else to monitor your application. 

> To my original question,

What if you created your own Runtime used only in app_dev.php and on the [`getRunner()`](https://github.com/symfony/symfony/blob/6.2/src/Symfony/Component/Runtime/SymfonyRuntime.php#L133) method you did your logic to know if you should return a 
```php
new ResponseRunner(new RedirectResponse('some homepage'))
```

??

**Reply by freiondrej-lmc** [reply]
@Nyholm thank you for your insights, they are really useful! 

As to the custom Runtime - that sounds really good, I suppose it could extend SymfonyRuntime to make things easier - something like 
```php
// CustomRuntime.php
public function getRunner(?object $application) {
   if ($customLogic) {
      return parent::getRunner(new RedirectResponse('some homepage'));
   }
   return parent::getRunner($application);
}
```
What is the best way to specify that the custom runtime should be used? Something like:
```php
$_ENV['APP_RUNTIME'] = 'Namespace\CustomRuntime';

require_once __DIR__ . '/../vendor/autoload_runtime.php';

```
?

**Reply by freiondrej-lmc** [reply]
@Nyholm I implemented the solution as you suggested (I used the `$_ENV['APP_RUNTIME'] = 'Namespace\CustomRuntime';` to set the custom runtime) and the code is much nicer and it works like a charm. Thank you for this amazing component! 🚀
